File: app/src/main/java/com/example/recipeclipper/data/local/entity/Entities.kt
package com.example.recipeclipper.data.local.entity

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey
import java.util.UUID

@Entity(
    tableName = "recipes",
    indices = [Index(value = ["sourceUrl"], unique = true), Index(value = ["uid"], unique = true)]
)
data class RecipeEntity(
    @PrimaryKey(autoGenerate = true) val id: Long = 0,
    val sourceUrl: String,          // unique: re-sharing a link upserts instead of duplicating
    val title: String,
    val imageUrl: String?,
    val ingredients: List<String>,  // TypeConverter (JSON)
    val instructions: List<String>,
    val prepTime: String?,
    val cookTime: String?,
    val totalTime: String?,
    val servings: String?,          // the recipe's yield text as published
    val sourceType: String,         // BLOG | REDDIT, for re-fetch
    val lastViewedAt: Long,
    val checkedIngredients: Set<Int> = emptySet(),
    val notes: String? = null,      // the user's own note; kept across re-shares
    /** Stable across devices and exports (#26): what an export file calls this recipe. Never
     *  changes once written — re-sharing keeps it, and an import keeps the file's. */
    @ColumnInfo(defaultValue = "")
    val uid: String = newUid(),
    val language: String? = null,   // the recipe's language tag (#14); null before version 5
    val cookState: String? = null,  // CookProgress as JSON (CookStateJson); kept if steps unchanged
    val servingsTarget: Int? = null, // the chosen servings; null = the recipe's own yield
    /** [com.example.recipeclipper.data.model.ContentOrigin] by name (#29): PARSED is the
     *  source's words, anything else the user's version, never refreshed by a re-share. */
    @ColumnInfo(defaultValue = "PARSED")
    val contentOrigin: String = "PARSED",
    val editedAt: Long? = null      // when the user last saved an edit; null if never
)

@Entity(tableName = "lists", indices = [Index(value = ["uid"], unique = true)])
data class ListEntity(
    @PrimaryKey(autoGenerate = true) val id: Long = 0,
    val name: String,
    // Seeded on first run, and sorts before user-created lists. Deliberately NOT "undeletable"
    // — that is isFavorites' job, and only Favorites is protected.
    val isBuiltIn: Boolean,
    val isFavorites: Boolean,       // exactly one row; a column, never a name match
    val sortOrder: Int,
    val createdAt: Long,
    /** Stable across devices and exports (#26); survives a rename, unlike [name]. */
    @ColumnInfo(defaultValue = "")
    val uid: String = newUid()
)

/**
 * One line on the grocery list (#50). [text] is the line as written: a recipe's line as the
 * reading view showed it (scaled and converted), or what was typed. [language] is the tag
 * whose words read it (null: none, so it is never named or combined). [aisle] is an `Aisle`
 * key, chosen from the aisle table when added and changed only by the user. [listId] is
 * [DEFAULT_LIST] until there are several lists: a `grocery_lists` table can come later with
 * this as its key. A recipe's items stay when it is deleted; they just lose their source
 * (SET NULL). [uid] and [updatedAt] are for export and a later sync (#53).
 */
@Entity(
    tableName = "grocery_items",
    foreignKeys = [
        ForeignKey(
            entity = RecipeEntity::class,
            parentColumns = ["id"],
            childColumns = ["recipeId"],
            onDelete = ForeignKey.SET_NULL
        )
    ],
    indices = [
        Index(value = ["uid"], unique = true),
        Index("listId"),
        Index("recipeId")
    ]
)
data class GroceryItemEntity(
    @PrimaryKey(autoGenerate = true) val id: Long = 0,
    val listId: Long = DEFAULT_LIST,
    val text: String,
    val language: String?,
    val aisle: String,
    val checked: Boolean = false,
    val sortOrder: Int,             // the order added, within the list
    val recipeId: Long?,            // the recipe it came from, if any
    val plannedDay: Long?,          // the planned day it came from (an epoch day), if any
    val updatedAt: Long,
    val uid: String = newUid()
) {
    companion object {
        const val DEFAULT_LIST = 1L
    }
}

/**
 * One pantry item (#51). [name] as typed, read with [language]'s words; [quantity] free text
 * as written; [aisle] an `Aisle` key, chosen when added. [purchasedDay] and [expiresDay] are
 * epoch days. [alwaysHave] marks a staple. No foreign keys: an item belongs to no recipe.
 * [uid] and [updatedAt] are for export and a later sync (#53).
 */
@Entity(tableName = "pantry_items", indices = [Index(value = ["uid"], unique = true)])
data class PantryItemEntity(
    @PrimaryKey(autoGenerate = true) val id: Long = 0,
    val name: String,
    val quantity: String?,
    val language: String?,
    val aisle: String,
    val inStock: Boolean,
    val alwaysHave: Boolean,
    val purchasedDay: Long?,
    val expiresDay: Long?,
    val updatedAt: Long,
    val uid: String = newUid()
)

/** A fresh stable id for a new row. The same form the migrations backfill with. */
fun newUid(): String = UUID.randomUUID().toString()

/** "Saved" means "has at least one of these". There is deliberately no isSaved column. */
@Entity(
    tableName = "recipe_list_cross_ref",
    primaryKeys = ["recipeId", "listId"],
    foreignKeys = [
        ForeignKey(
            entity = RecipeEntity::class,
            parentColumns = ["id"],
            childColumns = ["recipeId"],
            onDelete = ForeignKey.CASCADE
        ),
        ForeignKey(
            entity = ListEntity::class,
            parentColumns = ["id"],
            childColumns = ["listId"],
            onDelete = ForeignKey.CASCADE
        )
    ],
    indices = [Index("listId")]
)
data class RecipeListCrossRef(
    val recipeId: Long,
    val listId: Long,
    val addedAt: Long
)

/**
 * A meal type (#49). Seeded rows carry a [builtInKey] ("breakfast", "lunch", "dinner",
 * "snack") that survives a rename; only rows without one can be deleted, a guard kept in the
 * SQL. [uid] and [updatedAt] are for export and a later sync (#53).
 */
@Entity(tableName = "meal_types", indices = [Index(value = ["uid"], unique = true)])
data class MealTypeEntity(
    @PrimaryKey(autoGenerate = true) val id: Long = 0,
    val name: String,
    val builtInKey: String?,
    val sortOrder: Int,
    val updatedAt: Long,
    val uid: String = newUid()
)

/**
 * One meal on the plan (#49): a recipe with its planned servings, or a note, on a local
 * calendar [day] (an epoch day, see `PlanDays`). A recipe's entries go with it when it is
 * deleted (cascade). A meal type can't be deleted from under its entries (no cascade):
 * deleting one first moves them to Dinner, in the same transaction.
 */
@Entity(
    tableName = "meal_plan_entries",
    foreignKeys = [
        ForeignKey(
            entity = MealTypeEntity::class,
            parentColumns = ["id"],
            childColumns = ["mealTypeId"]
        ),
        ForeignKey(
            entity = RecipeEntity::class,
            parentColumns = ["id"],
            childColumns = ["recipeId"],
            onDelete = ForeignKey.CASCADE
        )
    ],
    indices = [
        Index(value = ["uid"], unique = true),
        Index("day"),
        Index("mealTypeId"),
        Index("recipeId")
    ]
)
data class MealPlanEntryEntity(
    @PrimaryKey(autoGenerate = true) val id: Long = 0,
    val day: Long,
    val mealTypeId: Long,
    val recipeId: Long?,
    val servings: Int?,             // planned servings; null = the recipe's own yield
    val note: String?,              // a note instead of a recipe
    val sortOrder: Int,             // within its day and meal type
    val updatedAt: Long,
    val uid: String = newUid()
)

/**
 * A reusable weekly menu (#52): a named copy of a week's meals, to add to any later week. Its
 * entries live in [MenuEntryEntity]. [uid] and [updatedAt] are for export and a later sync (#53).
 */
@Entity(tableName = "menus", indices = [Index(value = ["uid"], unique = true)])
data class MenuEntity(
    @PrimaryKey(autoGenerate = true) val id: Long = 0,
    val name: String,
    val updatedAt: Long,
    val uid: String = newUid()
)

/**
 * One meal of a menu (#52), on [dayOffset] days after the week's first day (0 to 6), shaped like
 * [MealPlanEntryEntity]: a recipe at [servings] (null: its own yield), or a [note]. Entries go
 * with their menu and with their recipe (cascade). A meal type can't be deleted from under them
 * (no cascade): deleting one moves them to Dinner, as it does planned meals.
 */
@Entity(
    tableName = "menu_entries",
    foreignKeys = [
        ForeignKey(
            entity = MenuEntity::class,
            parentColumns = ["id"],
            childColumns = ["menuId"],
            onDelete = ForeignKey.CASCADE
        ),
        ForeignKey(
            entity = MealTypeEntity::class,
            parentColumns = ["id"],
            childColumns = ["mealTypeId"]
        ),
        ForeignKey(
            entity = RecipeEntity::class,
            parentColumns = ["id"],
            childColumns = ["recipeId"],
            onDelete = ForeignKey.CASCADE
        )
    ],
    indices = [
        Index(value = ["uid"], unique = true),
        Index("menuId"),
        Index("mealTypeId"),
        Index("recipeId")
    ]
)
data class MenuEntryEntity(
    @PrimaryKey(autoGenerate = true) val id: Long = 0,
    val menuId: Long,
    val dayOffset: Int,
    val mealTypeId: Long,
    val recipeId: Long?,
    val servings: Int?,
    val note: String?,
    val sortOrder: Int,             // within its day and meal type
    val updatedAt: Long,
    val uid: String = newUid()
)

/**
 * Chef mode's short version of one step (#100), written on the device. Derived data, never in
 * the export file: keyed by recipe, the step's text ([stepHash], its SHA-256) and the language
 * it was written in, so a changed step has no row and is written again; rows for steps the
 * recipe no longer has are pruned. [shortText] null: the model's version failed
 * `ShortStepCheck`, so the step shows as written and isn't asked for again. [uid] and
 * [updatedAt] follow the other tables (#53).
 */
@Entity(
    tableName = "short_steps",
    foreignKeys = [
        ForeignKey(
            entity = RecipeEntity::class,
            parentColumns = ["id"],
            childColumns = ["recipeId"],
            onDelete = ForeignKey.CASCADE
        )
    ],
    indices = [
        Index(value = ["uid"], unique = true),
        Index(value = ["recipeId", "stepHash", "language"], unique = true)
    ]
)
data class ShortStepEntity(
    @PrimaryKey(autoGenerate = true) val id: Long = 0,
    val recipeId: Long,
    val stepHash: String,
    val language: String,
    val shortText: String?,
    val updatedAt: Long,
    val uid: String = newUid()
)

/**
 * The on-device model's typed decisions (#104), one row per question: its [kind] key, its
 * [input] normalised and its [language], and the [answer] after the confidence rule (an option
 * or "unsure"), so each question is asked once. Derived data: never exported. [uid] and
 * [updatedAt] follow the other tables (#53).
 */
@Entity(
    tableName = "ai_decisions",
    indices = [
        Index(value = ["uid"], unique = true),
        Index(value = ["kind", "input", "language"], unique = true)
    ]
)
data class AiDecisionEntity(
    @PrimaryKey(autoGenerate = true) val id: Long = 0,
    val kind: String,
    val input: String,
    val language: String,
    val answer: String,
    val updatedAt: Long,
    val uid: String = newUid()
)

File: ios/RecipeClipper/Data/Local/Entities.swift
import Foundation

// Database rows, distinct from the domain types in Model/Recipe.swift (the repositories map
// between them in RecipeMapping.swift). Mirrors Android's RecipeEntity / ListRow /
// RecipeSummaryRow. Cross-ref rows use the domain `ListMembership` directly: its three fields
// are exactly the table's three columns.

/// One row of `recipes`.
struct RecipeRecord: Equatable {
    var id: Int64 = 0                 // 0 = not yet inserted; SQLite assigns one
    var sourceUrl: String             // unique: re-sharing a link upserts instead of duplicating
    var title: String
    var imageUrl: String?
    var ingredients: [String]         // JSON text
    var instructions: [String]        // JSON text
    var prepTime: String?
    var cookTime: String?
    var totalTime: String?
    var servings: String?             // the recipe's yield text as published
    var sourceType: String            // BLOG | REDDIT, for re-fetch
    var lastViewedAt: Int64
    var checkedIngredients: Set<Int> = []   // JSON text, sorted
    var notes: String? = nil                // the user's own note; kept across re-shares
    /// Stable across devices and exports (#26): what an export file calls this recipe. Never
    /// changes once written — `update` doesn't touch it, and an import keeps the file's.
    var uid: String = newUid()
    var language: String? = nil             // the recipe's language tag (#14); nil before user_version 4
    var cookState: String? = nil            // CookProgress as JSON (CookStateJSON); kept if steps unchanged
    var servingsTarget: Int? = nil          // the chosen servings; nil = the recipe's own yield
    /// `ContentOrigin` by name (#29): PARSED is the source's words, anything else the user's
    /// version, never refreshed by a re-share.
    var contentOrigin: String = "PARSED"
    var editedAt: Int64? = nil              // when the user last saved an edit; nil if never

    /// The column list every `SELECT` of a full row uses, in `init(row:)`'s order.
    static let columns = """
        id, sourceUrl, title, imageUrl, ingredients, instructions, prepTime, cookTime, \
        totalTime, servings, sourceType, lastViewedAt, checkedIngredients, notes, uid, language, \
        cookState, servingsTarget, contentOrigin, editedAt
        """

    init(
        id: Int64 = 0, sourceUrl: String, title: String, imageUrl: String?,
        ingredients: [String], instructions: [String],
        prepTime: String?, cookTime: String?, totalTime: String?, servings: String?,
        sourceType: String, lastViewedAt: Int64, checkedIngredients: Set<Int> = [],
        notes: String? = nil, uid: String = newUid(), language: String? = nil,
        cookState: String? = nil, servingsTarget: Int? = nil,
        contentOrigin: String = "PARSED", editedAt: Int64? = nil
    ) {
        self.id = id
        self.sourceUrl = sourceUrl
        self.title = title
        self.imageUrl = imageUrl
        self.ingredients = ingredients
        self.instructions = instructions
        self.prepTime = prepTime
        self.cookTime = cookTime
        self.totalTime = totalTime
        self.servings = servings
        self.sourceType = sourceType
        self.lastViewedAt = lastViewedAt
        self.checkedIngredients = checkedIngredients
        self.notes = notes
        self.uid = uid
        self.language = language
        self.cookState = cookState
        self.servingsTarget = servingsTarget
        self.contentOrigin = contentOrigin
        self.editedAt = editedAt
    }

    init(row: SQLiteRow) {
        id = row.int64(0)
        sourceUrl = row.string(1)
        title = row.string(2)
        imageUrl = row.optionalString(3)
        ingredients = JSONColumns.decodeStrings(row.string(4))
        instructions = JSONColumns.decodeStrings(row.string(5))
        prepTime = row.optionalString(6)
        cookTime = row.optionalString(7)
        totalTime = row.optionalString(8)
        servings = row.optionalString(9)
        sourceType = row.string(10)
        lastViewedAt = row.int64(11)
        checkedIngredients = JSONColumns.decodeInts(row.string(12))
        notes = row.optionalString(13)
        uid = row.string(14)
        language = row.optionalString(15)
        cookState = row.optionalString(16)
        servingsTarget = row.isNull(17) ? nil : row.int(17)
        contentOrigin = row.string(18)
        editedAt = row.isNull(19) ? nil : row.int64(19)
    }
}

/// A fresh stable id for a new row: a lowercase UUID, the same form Android and the
/// migration's backfill use.
func newUid() -> String { UUID().uuidString.lowercased() }

/// One row of `meal_types` (#49; Android's `MealTypeEntity`).
struct MealTypeRecord: Equatable {
    var id: Int64 = 0
    var name: String
    var builtInKey: String?
    var sortOrder: Int
    var updatedAt: Int64
    var uid: String = newUid()

    static let columns = "id, name, builtInKey, sortOrder, updatedAt, uid"

    init(id: Int64 = 0, name: String, builtInKey: String?, sortOrder: Int, updatedAt: Int64, uid: String = newUid()) {
        self.id = id
        self.name = name
        self.builtInKey = builtInKey
        self.sortOrder = sortOrder
        self.updatedAt = updatedAt
        self.uid = uid
    }

    init(row: SQLiteRow) {
        id = row.int64(0)
        name = row.string(1)
        builtInKey = row.optionalString(2)
        sortOrder = row.int(3)
        updatedAt = row.int64(4)
        uid = row.string(5)
    }
}

/// One row of `meal_plan_entries` (#49; Android's `MealPlanEntryEntity`): a recipe with its
/// servings, or a note, on a local epoch `day`.
struct MealPlanEntryRecord: Equatable {
    var id: Int64 = 0
    var day: Int64
    var mealTypeId: Int64
    var recipeId: Int64?
    var servings: Int?              // planned servings; nil = the recipe's own yield
    var note: String?               // a note instead of a recipe
    var sortOrder: Int              // within its day and meal type
    var updatedAt: Int64
    var uid: String = newUid()

    static let columns = "id, day, mealTypeId, recipeId, servings, note, sortOrder, updatedAt, uid"

    init(
        id: Int64 = 0, day: Int64, mealTypeId: Int64, recipeId: Int64?, servings: Int?, note: String?,
        sortOrder: Int, updatedAt: Int64, uid: String = newUid()
    ) {
        self.id = id
        self.day = day
        self.mealTypeId = mealTypeId
        self.recipeId = recipeId
        self.servings = servings
        self.note = note
        self.sortOrder = sortOrder
        self.updatedAt = updatedAt
        self.uid = uid
    }

    init(row: SQLiteRow) {
        id = row.int64(0)
        day = row.int64(1)
        mealTypeId = row.int64(2)
        recipeId = row.isNull(3) ? nil : row.int64(3)
        servings = row.isNull(4) ? nil : row.int(4)
        note = row.optionalString(5)
        sortOrder = row.int(6)
        updatedAt = row.int64(7)
        uid = row.string(8)
    }
}

/// One line on the grocery list (#50; Android's `GroceryItemEntity`). `text` is the line as
/// written; `language` the tag whose words read it; `aisle` an `Aisle` key, chosen from the
/// aisle table when added and changed only by the user. `listId` is `defaultList` until there
/// are several lists. A recipe's items outlive it (SET NULL).
struct GroceryItemRecord: Equatable {
    var id: Int64 = 0
    var listId: Int64 = GroceryItemRecord.defaultList
    var text: String
    var language: String?
    var aisle: String
    var checked = false
    var sortOrder: Int
    var recipeId: Int64?
    var plannedDay: Int64?
    var updatedAt: Int64
    var uid: String = newUid()

    static let defaultList: Int64 = 1

    static let columns = "id, listId, text, language, aisle, checked, sortOrder, recipeId, plannedDay, updatedAt, uid"

    init(
        id: Int64 = 0, listId: Int64 = GroceryItemRecord.defaultList, text: String, language: String?, aisle: String,
        checked: Bool = false, sortOrder: Int, recipeId: Int64?, plannedDay: Int64?, updatedAt: Int64,
        uid: String = newUid()
    ) {
        self.id = id
        self.listId = listId
        self.text = text
        self.language = language
        self.aisle = aisle
        self.checked = checked
        self.sortOrder = sortOrder
        self.recipeId = recipeId
        self.plannedDay = plannedDay
        self.updatedAt = updatedAt
        self.uid = uid
    }

    init(row: SQLiteRow) {
        id = row.int64(0)
        listId = row.int64(1)
        text = row.string(2)
        language = row.optionalString(3)
        aisle = row.string(4)
        checked = row.bool(5)
        sortOrder = row.int(6)
        recipeId = row.isNull(7) ? nil : row.int64(7)
        plannedDay = row.isNull(8) ? nil : row.int64(8)
        updatedAt = row.int64(9)
        uid = row.string(10)
    }
}

/// One pantry item (#51; Android's `PantryItemEntity`).
struct PantryItemRecord: Equatable {
    var id: Int64 = 0
    var name: String
    var quantity: String?
    var language: String?
    var aisle: String
    var inStock: Bool
    var alwaysHave: Bool
    var purchasedDay: Int64?
    var expiresDay: Int64?
    var updatedAt: Int64
    var uid: String = newUid()

    static let columns = "id, name, quantity, language, aisle, inStock, alwaysHave, purchasedDay, expiresDay, updatedAt, uid"

    init(
        id: Int64 = 0, name: String, quantity: String?, language: String?, aisle: String, inStock: Bool,
        alwaysHave: Bool, purchasedDay: Int64?, expiresDay: Int64?, updatedAt: Int64, uid: String = newUid()
    ) {
        self.id = id
        self.name = name
        self.quantity = quantity
        self.language = language
        self.aisle = aisle
        self.inStock = inStock
        self.alwaysHave = alwaysHave
        self.purchasedDay = purchasedDay
        self.expiresDay = expiresDay
        self.updatedAt = updatedAt
        self.uid = uid
    }

    init(row: SQLiteRow) {
        id = row.int64(0)
        name = row.string(1)
        quantity = row.optionalString(2)
        language = row.optionalString(3)
        aisle = row.string(4)
        inStock = row.bool(5)
        alwaysHave = row.bool(6)
        purchasedDay = row.isNull(7) ? nil : row.int64(7)
        expiresDay = row.isNull(8) ? nil : row.int64(8)
        updatedAt = row.int64(9)
        uid = row.string(10)
    }

    var domain: PantryItem {
        PantryItem(
            id: id, name: name, quantity: quantity, language: language, aisle: Aisle.fromKey(aisle),
            inStock: inStock, alwaysHave: alwaysHave, purchasedDay: purchasedDay, expiresDay: expiresDay
        )
    }
}

/// A recipe's saved cook progress, with what a timer alert needs to name it (Android's
/// `CookStateRow`).
struct CookStateRecord: Equatable {
    let id: Int64
    let title: String
    let cookState: String
}

/// One row of a list of recipes: everything except the ingredients and steps.
struct RecipeSummaryRecord: Equatable {
    let id: Int64
    let title: String
    let imageUrl: String?
    let totalTime: String?
    let lastViewedAt: Int64
    let isSaved: Bool
    let isClipped: Bool
    let lastCookedDay: Int64?

    init(row: SQLiteRow) {
        id = row.int64(0)
        title = row.string(1)
        imageUrl = row.optionalString(2)
        totalTime = row.optionalString(3)
        lastViewedAt = row.int64(4)
        isSaved = row.bool(5)
        isClipped = row.bool(6)
        lastCookedDay = row.isNull(7) ? nil : row.int64(7)
    }
}

/// One list as a screen needs it. `containsRecipe` is always false when the query was asked
/// about `ListDao.noRecipe`.
struct ListRecord: Equatable {
    let id: Int64
    let name: String
    let isBuiltIn: Bool
    let isFavorites: Bool
    let recipeCount: Int
    let containsRecipe: Bool

    init(row: SQLiteRow) {
        id = row.int64(0)
        name = row.string(1)
        isBuiltIn = row.bool(2)
        isFavorites = row.bool(3)
        recipeCount = row.int(4)
        containsRecipe = row.bool(5)
    }
}

extension ListMembership {
    init(row: SQLiteRow) {
        self.init(recipeId: row.int64(0), listId: row.int64(1), addedAt: row.int64(2))
    }
}

/// The ingredient and step lists, and the ticked-ingredient set, are stored as JSON text in a
/// single column (Android's Converters). JSON rather than a join so a comma or quote in an
/// ingredient survives the round trip. Slashes are left unescaped so a search for "1/2" can
/// find "1/2 cup" in the stored ingredients column.
enum JSONColumns {
    static func encode(_ strings: [String]) -> String {
        encodeJSON(strings)
    }

    static func encode(_ ints: Set<Int>) -> String {
        encodeJSON(ints.sorted())
    }

    static func decodeStrings(_ json: String) -> [String] {
        (decodeJSON(json) as? [Any])?.compactMap { $0 as? String } ?? []
    }

    static func decodeInts(_ json: String) -> Set<Int> {
        Set((decodeJSON(json) as? [Any])?.compactMap { ($0 as? NSNumber)?.intValue } ?? [])
    }

    private static func encodeJSON(_ array: [Any]) -> String {
        guard let data = try? JSONSerialization.data(withJSONObject: array, options: [.withoutEscapingSlashes]),
              let text = String(data: data, encoding: .utf8)
        else { return "[]" }
        return text
    }

    private static func decodeJSON(_ json: String) -> Any? {
        guard let data = json.data(using: .utf8) else { return nil }
        return try? JSONSerialization.jsonObject(with: data)
    }
}
